pre_commit_all_the_way_down.py

```diff
@@ -0,0 +1,115 @@
+import argparse
+import re
+from pathlib import Path
+from re import Match
+from sys import exit
+from tempfile import TemporaryDirectory
+from textwrap import dedent, indent
+from typing import Optional, Sequence
+
+import sh
+
+BLOCK_TYPES = "(code|code-block|sourcecode|ipython)"
+PY_LANGS = "(python|py|sage|python3|py3|numpy)"
+DOCTEST_TYPES = "(testsetup|testcleanup|testcode)"
+RST_RE = re.compile(
+    rf"(?P<before>"
+    rf"^(?P<indent> *)\.\. ("
+    rf"jupyter-execute::|"
+    rf"python-script::|"
+    rf"{BLOCK_TYPES}:: {PY_LANGS}|"
+    rf"{DOCTEST_TYPES}::.*"
+    rf")\n"
+    rf"((?P=indent) +:.*\n)*"
+    rf"\n*"
+    rf")"
+    rf"(?P<code>(^((?P=indent) +.*)?\n)+)",
+    re.MULTILINE,
+)
+INDENT_RE = re.compile("^ +(?=[^ ])", re.MULTILINE)
```

Comment:
Same remark as above. This isn't a non-capturing group, but a look-ahead assertion. It only matches if the spaces aren't followed by a space. I think this could be optimized, since regexps are greedy (so they _already_ consume all the whitespaces...)

Reply:
I got to pump up my regexp game